### PR TITLE
service/soc_u: Implement getaddrinfo and getnameinfo

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
Implements two additional functions in soc:U, getaddrinfo and getnameinfo, using the corresponding host socket functions.

I only tested this with a few (self-written) homebrews and things seem to work fine so far... more testing and reviews would be appreciated.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4632)

### CODE REVIEW

File: src/core/hle/service/soc_u.cpp [hunk 1]
```diff
@@ -858,6 +896,95 @@ void SOC_U::SetSockOpt(Kernel::HLERequestContext& ctx) {
...
+    rb.Push(count);
+    rb.PushStaticBuffer(out_buff, 0);
+}
+
+void SOC_U::GetNameInfo(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x10, 4, 2);
+    u32 socklen = rp.Pop<u32>();
+    u32 hostlen = rp.Pop<u32>();
+    u32 servlen = rp.Pop<u32>();
+    int flags = static_cast<int>(rp.Pop<u32>());
```
Comment (wwylele) [inline src/core/hle/service/soc_u.cpp]: - It is weird to have flags (as in bit flags?) to have a signed type
 - If this is indeed a signed type, Use `rp.Pop<s32>` instead

Comment (zhaowenlan1779) [reply src/core/hle/service/soc_u.cpp]: The corresponding POSIX function accepts this parameter as type `int`, so I guess this is actually signed?

Comment (wwylele) [reply src/core/hle/service/soc_u.cpp]: Then use `s32 flags = rp.Pop<s32>()` to explicitly express the underlying storage type

File: src/core/hle/service/soc_u.cpp [hunk 2]
```diff
@@ -51,6 +51,14 @@
 #define closesocket(x) close(x)
 #endif
 
+// Some platforms seem to have these defined, they conflict with our function names
+#ifdef GetAddrInfo
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: You may want to just consider naming the functions something else like GetAddrInfoImpl/GetNameInfoImpl, which avoids the need for platform hackery like this.

File: src/core/hle/service/soc_u.cpp [hunk 3]
```diff
@@ -320,6 +328,36 @@ union CTRSockAddr {
...
+    s32 ai_family;
+    s32 ai_socktype;
+    s32 ai_protocol;
+    s32 ai_addrlen;
+    char ai_canonname[256];
+    CTRSockAddr ai_addr;
+
+    /// Converts a platform-specific addrinfo to a 3ds addrinfo.
+    static CTRAddrInfo FromPlatform(const addrinfo& addr) {
+        CTRAddrInfo ctr_addr;
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]:
```suggestion
        CTRAddrInfo ctr_addr{};
```

Is an equivalent that doesn't require memset (yes all padding bits are zeroed out as well).

**Pull request merged**